File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null) {
            return list2;
        } else if (list2 == null) {
            return list1;
        }

        ListNode listA = new ListNode(list1.val, list1.next);
        ListNode listB = new ListNode(list2.val, list2.next);
        ListNode output = new ListNode(-1);
        ListNode dummy = output;
        while (listA != null && listB != null) {
            if (listA.val < listB.val) {
                dummy.next = listA;
                listA = listA.next;
            } else {
                dummy.next = listB;
                listB = listB.next;
            }
            dummy = dummy.next;
        }
        if (listA != null) {
            dummy.next = listA;
        } else {
            dummy.next = listB;
        }
        return output.next;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        class NodeWrapper:
            def __init__(self, node):
                self.val = node.val
                self.next = node.next
            
            # needed to do heap operations
            def __lt__(self, other):
                return self.val < other.val

        heap = []

        for linked_list in lists:
            if linked_list: heapq.heappush(heap, NodeWrapper(linked_list))
        
        ans = dummy = ListNode(-1)

        while heap:
            node = heapq.heappop(heap)
            dummy.next = ListNode(node.val)
            dummy = dummy.next
            if node.next:
                heapq.heappush(heap, NodeWrapper(node.next))

        return ans.next